### ISSUE TITLE
Calling browse with an obviously invalid argument should raise an error

### ISSUE THREAD

**Issue by LeartS** [odoo/odoo#12519]
**Impacted versions:**
8, didn't test with 9

**Steps to reproduce:**
* Call `somemodel.browse()` with invalid ids, for example `self.env['res.country'].browse([('id', '=', 2)])` which could happen if by mistake you call `browse` instead of `search`
* Before accessing a field of the recordset, call `samemodel.search` or `samemodel.browse` with correct parameters
* Try to access a field of the recordset, thus forcing fetching from the DB

**Current behavior:**

You get something similar to this:

```python
ipdb> !c = countries[0]
ipdb> p c.name
2016-06-21 12:59:11,929 27311 ERROR dev-db openerp.sql_db: bad query:  SELECT "res_country"."id","res_country"."write_uid","res_country"."name","res_country"."create_date","res_country"."default_language_id","res_country"."currency_id","res_country"."address_format","res_country"."write_date","res_country"."code","res_country"."create_uid" FROM "res_country"
                        WHERE "res_country".id IN (1, 134, 138, 87, 12, 13, 143, 16, 21, 150, 89, 153, 27, 31, 166, ('id', '=', 235), 181, 238, 183, 185, 58, 189, 65, 69, 71, 201, 74, 76, 79, 141, 83, 215, 217, 90, 92, 186, 102, 104, 233, 106, 235, 110, 111, 240, 241, 242, 248, 234, 149)  ORDER BY "res_country"."name"  
                    
Traceback (most recent call last):
  File "/opt/odoo-buildout/itmyit-dev/parts/odoo/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block
```

You can see the `('id', '=', 235)` in the sea of correct ids, which is the cause of the problem. The environment prefetch dict is a defaultdict so everything you pass to `browse` will be saved in the set of "ids" to prefetch

**This can be difficult to spot and debug (I've just spent a nice hour doing it) because the invalid `browse` could have been called in a completely different part of the code from where the fields are accessed and the error raised**, since the query is not executed until a field is accessed.

**Expected behavior:**

Odoo should give an error or at least output a warning in the log when `browse` is called with something that isn't an iterable of integers.

cc @rco-odoo

**Comment by xmo-odoo** [comment]
> Odoo should give an error or at least output a warning in the log when browse is called with something that isn't an iterable of integers.

A deep validation of the arguments was implemented early on, its impact on overall performance was too large to be acceptable (relative to its effective utility over the odoo codebase, there are *lots* of implicit browse calls the old/new API interface) and lead to its removal.

What *could* happen is an alternate and deep-validating implementation of `_normalize_ids` being used when running in a debug mode of some sort, though that'd have to be a different proposal, and likely for v10 not v8.

**Comment by LeartS** [comment]
Thanks for the answer, I guess it makes sense considering that often the list of ids passed to a query is relatively large. The debug-only check seems a nice compromise.
And I don't think a deep validation is really necessary, a simple

```
try:
    sum(ids)
except TypeError:
    logger.warn('Invalid ids supplied')
```

Could catch most cases, and takes ~10 microseconds for a list of 1000 elements

**Comment by xmo-odoo** [comment]
> Could catch most cases, and takes ~10 microseconds for a list of 1000 elements

1. that would *triple* the existing runtime of _normalize_ids, some of the optimisations in that method (e.g. swapping out `isinstance`) were considered worth implementing at significantly lower gain than that
2. would behave very oddly (and slowly) on string ids and would not work for NewId

**Comment by LeartS** [comment]
Sorry, my previous messages were ambiguous and I didn't make the point/my problem clear: if it's not possible or convenient to prevent the error, I think it's important to suggest what the source of the error could be to facilitate debug.

We are now discussing after having confirmed that the cause of the problem was an incorrect `browse` call, but for someone who is developing normally and finds himself with an error like the one I posted, it's not immediately clear that the source of the "strange id" in the list is (probably) a call to browse with that value.

That's because **it's not obvious** that whatever you pass to `browse`, be it a int, list, a `search` domain, a JSON file as a python dict, or anything that can be passed to `tuple()`, ends up in the set of ids to prefetch at the next query.

So developers, like me, can find themeselves asking "how did that thing end there?" and will probably assume a bug somewhere in the deep part of odoo code and waste hours debugging.

With that said, my proposal is not something to be checked at each browse call, but immediately before a query is launched; or even only after it executed, and only if it failed. In these cases there shouldn't be any string ids or NewIds and a few microseconds is negligible compared to the query latency.

**TL;DR:** alert users: "There was an error in the query, if there are strange ids, keep in mind that whatever you pass to `browse` is put in the `WHERE` clause as-is and that's probably the cause of the problem"
If not an error/warning message in the logs, at least make it clear in the documentation.
